daemon/src/collab_settlement_maker.rs

```diff
@@ -104,100 +105,111 @@ impl xtra::Actor for Actor {
-                e
-            );
-        };
+    async fn handle_proposal(&mut self) -> Result<()> {
+        let mut conn = self.db.acquire().await?;
+        let cfd = load_cfd(self.proposal.order_id, &mut conn).await?;
+
+        let event = cfd.receive_collaborative_settlement_proposal(self.proposal.clone())?;
+        self.process_manager
+            .send(process_manager::Event::new(event))
+            .await??;
+
+        Ok(())
     }
 
     async fn complete(
         &mut self,
         completed: Completed<CollaborativeSettlement>,
         ctx: &mut xtra::Context<Self>,
     ) {
-        let _ = self
-            .on_completed
-            .send(completed)
-            .log_failure("Failed to inform about collab settlement completion")
-            .await;
+        let order_id = self.proposal.order_id;
+        let event_fut = async {
+            let mut conn = self.db.acquire().await?;
+            let cfd = load_cfd(order_id, &mut conn).await?;
+            let event = cfd.settle_collaboratively(completed)?;
+
+            anyhow::Ok(event)
+        };
+
+        match event_fut.await {
+            Ok(event) => {
+                let _ = self
+                    .process_manager
+                    .send(process_manager::Event::new(event))
+                    .await;
+            }
+            Err(e) => {
+                tracing::warn!(%order_id, "Failed to report completion of collab settlement: {:#}", e)
+            }
+        }
 
         ctx.stop();
     }
 
-    async fn accept(&mut self, ctx: &mut xtra::Context<Self>) {
-        let this = ctx.address().expect("self to be alive");
-        self.inform_taker(
-            maker_inc_connections::settlement::Decision::Accept { address: this },
-            ctx,
-        )
-        .await
-    }
+    async fn accept(&mut self, ctx: &mut xtra::Context<Self>) -> Result<()> {
+        let order_id = self.proposal.order_id;
+        tracing::info!(%order_id,
+                       "Settlement proposal accepted");
 
-    async fn reject(&mut self, ctx: &mut xtra::Context<Self>) {
-        self.inform_taker(maker_inc_connections::settlement::Decision::Reject, ctx)
+        let mut conn = self.db.acquire().await?;
+        let cfd = load_cfd(order_id, &mut conn).await?;
+        let event = cfd.accept_collaborative_settlement_proposal(&self.proposal)?;
+
+        self.process_manager
+            .send(process_manager::Event::new(event))
+            .await??;
+
+        let this = ctx.address().expect("self to be alive");
+        self.connections
+            .send(maker_inc_connections::settlement::Response {
+                taker_id: self.taker_id,
+                order_id,
+                decision: maker_inc_connections::settlement::Decision::Accept { address: this },
+            })
             .await
+            .context("Failed to inform taker about settlement acceptance")??;
```

Comment:
Good point, but I think I still prefer my approach here.

If we fail to send this message to the taker, we must `complete` the collaborative settlement protocol with `Failed`, much like with the rest of the errors in this method. Keep in mind that this `accept` method is called by `handle(&mut self, msg: Accepted)`, where any error originated from this results in completing the protocol with `Failed`.

I did reject differently because it is simpler to complete the protocol with `Rejected` unconditionally. If the rejection message doesn't reach the taker, I don't see a clear upside to recording the fact that it failed (whilst it was being rejected).